### DISCUSSION TITLE
Save videos repeatedly during experiment

### DISCUSSION THREAD

**Discussion by roaldarbol** [bonsai-rx/bonsai#1597] [Q&A] [answered]
Just continuing the conversation from Discord. 

> I'm recording 24h videos from 2 cameras using VideoCapture and VideoWriter, but ran out of memory after ~6h. Is there a way I can save the files every X hours/minutes? I'm thinking some combination of Timer and Repeat, but it's not quite intuitive to me how I'll achieve that.

@bruno-f-cruz Would be really happy to see get some more information about how you would go about it with CreateObservable->Switch. Thanks for the help!

**Comment by bruno-f-cruz** [comment, marked as answer]
This is what I had in mind. The timer restarts an observable that subscribes to the video every N minutes.

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/9157339a-9e4d-4add-8c29-9ee15372497c)

<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml
<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.8.1"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:cv="clr-namespace:Bonsai.Vision;assembly=Bonsai.Vision"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="cv:CameraCapture">
          <cv:Index>0</cv:Index>
          <cv:CaptureProperties />
        </Combinator>
      </Expression>
      <Expression xsi:type="rx:PublishSubject">
        <Name>video</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Timer">
          <rx:DueTime>PT0S</rx:DueTime>
          <rx:Period>PT1M</rx:Period>
        </Combinator>
      </Expression>
      <Expression xsi:type="rx:CreateObservable">
        <Workflow>
          <Nodes>
            <Expression xsi:type="SubscribeSubject">
              <Name>video</Name>
            </Expression>
            <Expression xsi:type="Combinator">
              <Combinator xsi:type="cv:VideoWriter">
                <cv:FileName>video.avi</cv:FileName>
                <cv:Suffix>Timestamp</cv:Suffix>
                <cv:Buffered>true</cv:Buffered>
                <cv:Overwrite>false</cv:Overwrite>
                <cv:FourCC>FMP4</cv:FourCC>
                <cv:FrameRate>30</cv:FrameRate>
                <cv:FrameSize>
                  <cv:Width>0</cv:Width>
                  <cv:Height>0</cv:Height>
                </cv:FrameSize>
                <cv:ResizeInterpolation>NearestNeighbor</cv:ResizeInterpolation>
              </Combinator>
            </Expression>
          </Nodes>
          <Edges>
            <Edge From="0" To="1" Label="Source1" />
          </Edges>
        </Workflow>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Switch" />
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="2" To="3" Label="Source1" />
      <Edge From="3" To="4" Label="Source1" />
    </Edges>
  </Workflow>
</WorkflowBuilder>
```

</details>

**Reply by roaldarbol** [reply]
Thanks an bunch! I can see how this approach can be helpful in quite a few situations! Unfortunately this doesn't work for me - I can make the timer tick (and see that plotted) and see the video (both using CamperaCapture and VideoCaptureDevice), but nothing is saving. I don't have any experience with the Observables, so I hope you can help me on here. I did try changing the timings in the timer too (trying both shorter and longer as well as a bunch of combinations) but no luck unfortunately. 

EDIT Here's a screenshot of what everything looks like:
![bonsai-screenshot](https://github.com/bonsai-rx/bonsai/assets/25629697/401194cd-f001-4bb8-9c13-017bddddbd3f)

**Reply by bruno-f-cruz** [reply]
Can you please attach the workflow? thanks
Oops, nevermind, I had an error in the workflow, can you give it another try?

**Reply by roaldarbol** [reply]
No worries! I'll give it a go in the morning again☺️

**Reply by roaldarbol** [reply]
Works beautifully now, thanks a ton! Was the issue that the Source>Workflow Output was doing something that interfered with the VideoWriter process?

**Reply by Joaqo98** [reply]
Hello! sorry that I step in, but I was wondering how could I change the workflow in order to save videos of a given time length with time intervals between them.
For example, if I wanted to record a first video of 1 hour, then wait another hour and automatically start recording another 1 hour video, what parameters or nodes should I chenge/add?
I find myself trying to achieve this through the "due time" and "period" of the Timer node, but I haven't managed to do what I said.
Any comment would be super useful!
Thank you very much

**Reply by bruno-f-cruz** [reply]
Add a `TakeUntil` triggered by a second timer inside the `CreateObservable`. The inside timer determines how long each video is recorded for, the outer timer determines how often the Video is recorded.

**Reply by Joaqo98** [reply]
![bonsai-rx help](https://github.com/bonsai-rx/bonsai/assets/148992233/5b8c291a-4490-40a2-a7f7-bf33c87b6bc1)
thank you very much for your quick response. you mean something like this?
the issue is I still can't record a, for instance, 10 seconds video, then stop recording for five seconds and, after those 5 seconds, start recording again another 10 seconds video.
When I set the outer timer to 5 seconds and the inside timer to 10 seconds, no video is recorded at all
